---
title: "质控报告"
date: "`r Sys.Date()`"
output: 
  html_document:
    theme: paper
    css: index.css
    includes:
      in_header: extLogo.html
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(Seurat)
library(kableExtra)
library(knitr)
library(scales)
library(dplyr)
library(ggplot2)
library(DT)
library(plotly)
library(shiny)
mycol = c("#F56867","#FEB915","#C798EE","#59BE86","#7495D3","#D1D1D1","#6D1A9C",
        "#15821E","#3A84E6","#70e014","#787878","#DB4C6C","#0430e0","#554236",
        "#AF5F3C","#ff7700","#e00417","#DAB370","#fcfc05","#268785","#ed1299", 
        "#09f9f5", "#246b93", "#cc8e12", "#d561dd", "#c93f00", "#ddd53e","#4aef7b",
        "#e86502", "#9ed84e", "#39ba30", "#6ad157", "#8249aa", "#99db27", "#e07233",
        "#ff523f","#ce2523", "#f7aa5d", "#cebb10", "#03827f", "#931635", "#373bbf", 
        "#a1ce4c", "#ef3bb6", "#d66551","#1a918f", "#ff66fc", "#2927c4", "#7149af" ,
        "#57e559" ,"#8e3af4" ,"#f9a270" ,"#22547f", "#db5e92","#edd05e", "#6f25e8", 
        "#0dbc21", "#280f7a", "#6373ed", "#5b910f" ,"#7b34c1" ,"#0cf29a" ,"#d80fc1", 
        "#51f59b")
```

# **<font color=black> </font>** { .tabset }
## **<font color=black size=5>概述</font>**
<br>
**<font color=black size=5>样品信息:</font>**
<br>
<br>
```{r,echo = FALSE}
data <- read.table(file = "metrics_summary.csv",header = T,sep = ",",check.names = F)[,c(1:3,19)]
a <- data[1,1];b <- data[1,2];c <- data[1,3];d <- data[1,4]
```
<div class="text-area-container">  
<div class="text-area">  
<span style="color: steelblue; font-size: 50px; font-weight: bold">`r a`</span>  
<span style="color: black; font-size: 15px;font-weight: bold">Estimated Number of Cells</span>  
</div>  
<div class="text-area">  
<span style="color: steelblue; font-size: 50px; font-weight: bold">`r d`</span>  
<span style="color: black; font-size: 15px;font-weight: bold">Median UMI counts per cell</span>  
</div>  
<div class="text-area">  
<span style="color: steelblue; font-size: 50px; font-weight: bold">`r c`</span>  
<span style="color: black; font-size: 15px;font-weight: bold">Median genes per cell</span>  
</div>  
<div class="text-area">  
<span style="color: steelblue; font-size: 50px; font-weight: bold">`r b`</span>  
<span style="color: black; font-size: 15px;font-weight: bold">Mean reads per cell</span>  
</div>  
</div>

<br>
<br>
```{r, echo = FALSE,warning=FALSE}
dt1 <- read.table(file = "Qual.stat",header = F,sep = "\t")
names(dt1) <- c("Type","Number","Percent(%)")
knitr::kable(x = dt1,format = 'html',table.attr = "style='width:100%;'",caption = "**表1 测序Reads统计**") %>% 
  kableExtra::kable_styling(position = "float_left",full_width = F,bootstrap_options = c("striped", "hover"),font_size = 16)

data <- as.data.frame(t(read.table(file = "metrics_summary.csv",header = T,sep = ",",check.names = F)[,10:16]))
dt2 <- data.frame(Type = rownames(data),"Percent(%)" = data$V1)
names(dt2) <- c("Type","Percent(%)")
knitr::kable(x = dt2,format = 'html',table.attr = "style='width:70%;'",caption = "**表2 数据比对**") %>% 
  kableExtra::kable_styling(position = "right",full_width = F,bootstrap_options = c("striped", "hover"),font_size = 16)
```
<details>
<summary>表1 note说明</summary>
<font size=1.5>Number of Reads：Reads总数；</font></br>
<font size=1.5>Valid Barcodes：包含有效Barcode的Reads数；</font></br>
<font size=1.5>Valid UMIs：包含有效UMIs的Reads数；</font></br>
<font size=1.5>Final Valid Reads：同时包含有效barcode与UMI的最终Reads数。</font></br>
</details>
<details>
<summary>表2 note说明</summary>
<font size=1.5>Reads Mapped to Genomes：比对到参考基因组上的Reads在总Reads中占的比例；</font></br>
<font size=1.5>Reads Mapped Confidently to Genome：比对到参考基因组并得到转录本GTF信息支持的Reads在总Reads中占的比例；</font></br>
<font size=1.5>Reads Mapped Confidently to Intergenic Regions：比对到基因间区域的Reads在总Reads中占的比例；</font></br>
<font size=1.5>Reads Mapped Confidently to Intronic Regions：比对到内含子区域的Reads在总Reads中占的比例；</font></br>
<font size=1.5>Reads Mapped Confidently to Exonic Regions：比对到外显子区域的Reads在总Reads中占的比例；</font></br>
<font size=1.5>Reads Mapped Confidently to Transcriptome：比对到已知参考转录本的Reads在总Reads中占的比例；</font></br>
<font size=1.5>Reads Mapped Antisense to Gene：比对到基因反义链上的Reads所占的比例。</font></br>
</details>


<table style="display:inline;">  
<tr>  
<td style="width:42%;height:100%;">
```{r, echo = FALSE}
data <- as.data.frame(t(read.table(file = "metrics_summary.csv",header = T,sep = ",",check.names = F)[,c(1:3,19,18,6,17)]))
dt3 <- data.frame(Type = rownames(data),Values = data$V1)
knitr::kable(x = dt3,format = 'html',digits = 2,table.attr = "style='width:85%;height:100%'",caption = "**表3 细胞统计**") %>% 
  kableExtra::kable_styling(position = "left",bootstrap_options = c("striped", "hover"),font_size = 16,full_width = T)
```
</td>


```{r, echo = FALSE,warning=FALSE}
data1 <- Read10X("filtered_feature_bc_matrix")
data2 <- Read10X("raw_feature_bc_matrix")
object1 <- CreateSeuratObject(data1)
object2 <- CreateSeuratObject(data2)

object1@meta.data %>% arrange(desc(nCount_RNA)) -> metadata1
object2@meta.data %>% arrange(desc(nCount_RNA)) -> metadata2
metadata1$Barcodes <- 1:nrow(metadata1)
metadata2$Barcodes <- 1:nrow(metadata2)
CellID <- ifelse(rownames(metadata2) %in% rownames(metadata1),"Cells","Background")
metadata2$CellID <- CellID
```

<td style="width:100%;height:100%;">
```{r fig.align="right", warning=FALSE,echo = FALSE,out.height="100%",out.width="100%"}
p <- ggplot(data = metadata2,aes(x = Barcodes,y = nCount_RNA,color = CellID)) + 
    geom_line(size = 1.2) +
    scale_x_log10(expand = c(0,0)) +
    scale_y_log10(expand = c(0,0)) +
    scale_color_manual(values = c("steelblue","grey"),
                       breaks = c("Cells","Background"),
                       labels = c("Cells","Background")) + 
    labs(x = "Barcodes",y = "UMI counts",title = "Barcode Rank Plot",color = "") +
    theme_light(base_rect_size = 2) +
    theme(axis.title.x = element_text(size = 14,face = 2,margin = margin(t = 0.1,unit = "cm")),
          axis.title.y = element_text(size = 14,face = 2,margin = margin(r = 0.1,unit = "cm")),
          axis.text.x = element_text(size = 12,face = 2,margin = margin(t = 0.1,unit = "cm")),
          axis.text.y = element_text(size = 12,face = 2,margin = margin(r = 0.1,unit = "cm")),
          panel.grid = element_line(size = 0.8),
          legend.text = element_text(size = 12,face = 2),
          legend.key.width = unit(8,"mm"),
          legend.key.height = unit(8,"mm"),
          axis.ticks.length = unit(0,"mm"),
          plot.title = element_text(hjust = 0.5,size = 14,face = 2),
          plot.margin = margin(t = 1,r = 0,b = 0,l = 2,unit = "cm"))
p
```
</td>
</tr>  
</table>

<details>
<summary>表3 note说明</summary>
<font size=1.5>Estimated Number of Cells：检测到的细胞数；</font></br>
<font size=1.5>Mean Reads per Cell：每个cell中Reads的平均数；</font></br>
<font size=1.5>Median Genes per Cell：每个cell中基因的中位数；</font></br>
<font size=1.5>Median UMI Counts per Cell：每个cell的UMI中位数；</font></br>
<font size=1.5>Total Genes Detected：基因总数；</font></br>
<font size=1.5>Sequencing Saturation：测序饱和度；</font></br>
<font size=1.5>Fraction Reads in Cells：过滤后细胞reads数占总reads数比例。</font></br>
</details>



## **<font color=black size=5>基因表达</font>**
##### **<font color=black>UMAP映射</font>**

```{r, echo=FALSE,warning=FALSE,fig.height=10,fig.width=10,fig.align="center",dpi=300}
object1@meta.data$Barcode <- rownames(object1@meta.data)
UMAP_df <- read.table(file = "umap_df.xls",header = T,sep = "\t",stringsAsFactors = F)
umi <- inner_join(object1@meta.data,UMAP_df,by = "Barcode")
umi <- na.omit(umi)

umap_umi <- ggplot(umi, aes(x = UMAP_1, y = UMAP_2))+
  geom_point(aes(color = nCount_RNA),size=1, shape=16)+
  labs(colour = "UMI counts",title = "UMAP Projection of Cells Colored by UMI Counts")+
  theme_bw(base_rect_size = 1.2)+
  scale_color_continuous(type = "viridis") +
  theme(plot.title = element_text(hjust = 0.5,size = 10),
        axis.title = element_text(size = 10),
        axis.text = element_text(size = 8),
        legend.text = element_text(size = 8),
        legend.title = element_text(size = 10),
        plot.margin = margin(t = 0.5,r = 0.5,b = 0.5,l = 0.5,unit = "cm"))

ncol1 <- names(table(UMAP_df$cluster))
ncol2 <- paste0(names(table(UMAP_df$cluster)),"-",table(UMAP_df$cluster),"cells")
ncoldf <- data.frame(cluster = ncol1,newcluster = ncol2)
UMAP_df$cluster <- as.character(UMAP_df$cluster)
umap_cluster <- left_join(UMAP_df,ncoldf,by = "cluster")
umap_cluster$cluster <- umap_cluster$newcluster
umap_cluster$cluster <- as.factor(umap_cluster$cluster)

umap_pic <- ggplot(umap_cluster, aes(x = UMAP_1, y = UMAP_2))+
  geom_point(aes(color = cluster),size=1, shape = 16)+
  scale_color_manual(values = mycol) + 
  labs(colour = "",title = "UMAP Projection of Cells Colored by Clustering")+
  theme_bw(base_size = 10,base_rect_size = 1.2)+
  theme(plot.title = element_text(hjust = 0.5,size = 10),
        axis.title = element_text(size = 10),
        axis.text = element_text(size = 8),
        legend.text = element_text(size = 8),
        plot.margin = margin(t = 0.5,r = 0,b = 0.5,l = 0.5,unit = "cm"))


fluidRow(
  column(6,ggplotly(umap_umi) %>% 
           layout(xaxis = list(showgrid = T,showline = T),
                  yaxis = list(showgrid = T,showline = T))),
  column(6,ggplotly(umap_pic) %>% 
           layout(xaxis = list(showgrid = T,showline = T),
                  yaxis = list(showgrid = T,showline = T)))
)

```

（左图）这里显示了每个细胞的UMI总数。UMI计数较大的细胞可能比UMI计数较少的细胞具有更高的RNA含量。轴对应于UMAP算法产生的二维嵌入。在这个空间中，相互接近的细胞对的基因表达谱比相距较远的细胞对更相似。

（右图）通过聚类算法将细胞分配给特定cluster。聚类将具有相似表达谱的细胞聚集在一起。轴对应于UMAP算法产生的二维嵌入。在这个空间中，相互接近的细胞对的基因表达谱比相距甚远的细胞对的基因表达谱更相似。UMAP图形，注重全局，弱化局部结构。在低纬空间保留了高维空间细胞间的关系，反映了细胞群潜在的拓扑结构，对于细胞轨迹推断有很好的的应用。

##### **<font color=black>Marker基因</font>**

```{r, echo = FALSE,warning=FALSE}
marker <- read.table(file = "marker_gene.csv",header = T,sep = ",",check.names = F)
mak <- marker[,c("gene","cluster","p_val_adj","p_val","avg_log2FC","pct.1","pct.2")]

datatable(mak,
          rownames = FALSE,
          extensions = "Buttons",
          #filter = 'bottom',
          options = list(dom = 'Bfrtip',buttons = list(list(extend = 'copy',
                                                            text = '复制'),
                                                       list(extend = "print",
                                                            text = "打印"),
                                                       list(extend = "collection",
                                                            buttons = c("csv","excel","pdf"),
                                                            text = "下载")),
                         search = list(regex = TRUE)),
          #container = sketch, # 设置表头格式
          class='stripe', # 鼠标悬停,整行突出显示
          caption = '表4 Top Genes By Cluster (Log2 fold-change, p-value)') %>%
  formatRound('avg_log2FC', 3) %>% # 小数点位数
  formatSignif("p_val_adj",digits = 3) %>%
  formatSignif("p_val",digits = 3)

```

上表显示了每个cluster的前10个差异表达基因。p值是假设检验后得到的原始值。p_val_adj是`bonferroni`多重检验校正后得到的校正后的P值。avg_log2FC是两组之间平均表达差异倍数的对数值。pct.1是第一组中检测到表达该基因的细胞所占的百分比，pct.2是第二组中检测到表达该基因的细胞所占的百分比。

##### **<font color=black>饱和度</font>**

```{r, echo=FALSE,warning=FALSE,fig.height=10,fig.width=10,fig.align="center",dpi=300}
satur <- read.table(file = "Saturation.xls",header = F,sep = "\t",check.names = F)
xaxis1 <- strsplit(x = satur[1,2],split = ",")[[1]]
yaxis1 <- strsplit(x = satur[2,2],split = ",")[[1]]
xaxis2 <- strsplit(x = satur[3,2],split = ",")[[1]]
yaxis2 <- strsplit(x = satur[4,2],split = ",")[[1]]

curveline1 <- data.frame(xaxis1,yaxis1)
curveline2 <- data.frame(xaxis2,yaxis2)

names(curveline1) <- c("Mean_Reads_per_Cell","Sequencing_Saturation")
names(curveline2) <- c("Mean_Reads_per_Cell","Median_Genes_per_Cell")

curveline1$Mean_Reads_per_Cell <- as.numeric(curveline1$Mean_Reads_per_Cell)
curveline1$Sequencing_Saturation <- as.numeric(curveline1$Sequencing_Saturation)

curveline2$Mean_Reads_per_Cell <- as.numeric(curveline2$Mean_Reads_per_Cell)
curveline2$Median_Genes_per_Cell <- as.numeric(curveline2$Median_Genes_per_Cell)

Saturation1 <- ggplot(curveline1, aes(x = Mean_Reads_per_Cell, y = Sequencing_Saturation))+
  geom_line(size = 1,color = "steelblue")+
  labs(x = "Mean Reads per Cell",
       y = "Sequencing Saturation",
       title = "Sequencing Saturation")+
  theme_bw(base_rect_size = 1.2) +
  scale_y_continuous(breaks = c(0,0.2,0.4,0.6,0.8,1),
                     labels = c(0,0.2,0.4,0.6,0.8,1),
                     limits = c(0,1)) +
  theme(plot.title = element_text(hjust = 0,size = 12),
        axis.title = element_text(size = 10),
        axis.text = element_text(size = 8),
        plot.margin = margin(t = 1,r = 1,b = 1,l = 1,unit = "cm"))


Saturation2 <- ggplot(curveline2, aes(x = Mean_Reads_per_Cell, y = Median_Genes_per_Cell))+
  geom_line(size = 1,color = "steelblue")+
  labs(x = "Mean Reads per Cell",
       y = "Median Genes per Cell",
       title = "Median Genes per Cell")+
  theme_bw(base_rect_size = 1.2) +
  theme(plot.title = element_text(hjust = 0,size = 12),
        axis.title = element_text(size = 10),
        axis.text = element_text(size = 8),
        plot.margin = margin(t = 1,r = 1,b = 1,l = 1,unit = "cm"))

fluidRow(
  column(6,ggplotly(Saturation1) %>% 
           layout(xaxis = list(showgrid = T,showline = T),
                  yaxis = list(showgrid = T,showline = T))),
  column(6,ggplotly(Saturation2) %>% 
           layout(xaxis = list(showgrid = T,showline = T),
                  yaxis = list(showgrid = T,showline = T)))
)
```

（左图）展示了此测序文库在不同测序深度下（通过随机取样获得）的测序饱和度，直到达到观察到的测序深度。X轴表示对文库的有效测序片段作不同比例的随机取样所得到的测序片段总数，Y轴对应下采样饱和度点。测序饱和度反映测序文库的复杂度和测序深度，当所有转录的mRNA都已测序时，接近1.0（100%）。

（右图）展示了每个细胞的中位数基因数量作为下采样测序深度（以每个细胞的平均读数衡量）的曲线，直到达到观察到的测序深度。X轴表示对文库的有效测序片段作不同比例的随机取样所计算得到的每个细胞平均度数，Y轴对应下采样每个细胞中位基因数目。



# **<font color=white>占位</font>**

------------------
<center> Copyright © 2022-2024 青岛百创智能制造技术有限公司 京ICP备10042835号-6 </center>
<center> 公司地址：山东省青岛市黄岛区保税港区上海路20号三号楼</center>
<center> Tel:400-600-3186 Fax:010-57045001 E-mail:tech@biomarker.com.cn [关于我们](http://www.bmkmanu.com/about) Ver:2.1.4</center>


