---
title: "质控报告"
date: "`r Sys.Date()`"
output: 
  html_document:
    #toc: true
    #toc_float: true
    theme: paper
    css: index.css

    includes:
      in_header: extLogo.html
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(kableExtra)
library(htmltools)
library(plotly)
library(shiny)
library(knitr)
col = c("#F56867","#FEB915","#C798EE","#59BE86","#7495D3","#D1D1D1","#6D1A9C",
        "#15821E","#3A84E6","#70e014","#787878","#DB4C6C","#0430e0","#554236",
        "#AF5F3C","#ff7700","#e00417","#DAB370","#fcfc05","#268785","#ed1299", 
        "#09f9f5", "#246b93", "#cc8e12", "#d561dd", "#c93f00", "#ddd53e","#4aef7b",
        "#e86502", "#9ed84e", "#39ba30", "#6ad157", "#8249aa", "#99db27", "#e07233",
        "#ff523f","#ce2523", "#f7aa5d", "#cebb10", "#03827f", "#931635", "#373bbf", 
        "#a1ce4c", "#ef3bb6", "#d66551","#1a918f", "#ff66fc", "#2927c4", "#7149af" ,
        "#57e559" ,"#8e3af4" ,"#f9a270" ,"#22547f", "#db5e92","#edd05e", "#6f25e8", 
        "#0dbc21", "#280f7a", "#6373ed", "#5b910f" ,"#7b34c1" ,"#0cf29a" ,"#d80fc1", 
        "#51f59b")
```

### **<font color=black>1 数据相关统计</font>**
```{r, echo = FALSE,warning=FALSE}
dt1 <- read.table(file = "Qual.stat",header = F,sep = "\t")
names(dt1) <- c("Type","Number","Percent(%)")
knitr::kable(x = dt1,format = 'html',table.attr = "style='width:100%;'",caption = "**表1 测序Reads统计**") %>% 
  kableExtra::kable_styling(position = "float_left",full_width = F,bootstrap_options = c("striped", "hover"),font_size = 16)

data <- as.data.frame(t(read.table(file = "metrics_summary.csv",header = T,sep = ",",check.names = F)[,10:16]))
dt2 <- data.frame(Type = rownames(data),"Percent(%)" = data$V1)
names(dt2) <- c("Type","Percent(%)")
knitr::kable(x = dt2,format = 'html',table.attr = "style='width:70%;'",caption = "**表2 数据比对**") %>% 
  kableExtra::kable_styling(position = "right",full_width = F,bootstrap_options = c("striped", "hover"),font_size = 16)
```
<details>
<summary>表1 note说明</summary>
<font size=1.5>Number of Reads：Reads总数；</font></br>
<font size=1.5>Valid Barcodes：包含有效Barcode的Reads数；</font></br>
<font size=1.5>Valid UMIs：包含有效UMIs的Reads数；</font></br>
<font size=1.5>Final Valid Reads：同时包含有效barcode与UMI的最终Reads数。</font></br>
</details>
<details>
<summary>表2 note说明</summary>
<font size=1.5>Reads Mapped to Genomes：比对到参考基因组上的Reads在总Reads中占的比例；</font></br>
<font size=1.5>Reads Mapped Confidently to Genome：比对到参考基因组并得到转录本GTF信息支持的Reads在总Reads中占的比例；</font></br>
<font size=1.5>Reads Mapped Confidently to Intergenic Regions：比对到基因间区域的Reads在总Reads中占的比例；</font></br>
<font size=1.5>Reads Mapped Confidently to Intronic Regions：比对到内含子区域的Reads在总Reads中占的比例；</font></br>
<font size=1.5>Reads Mapped Confidently to Exonic Regions：比对到外显子区域的Reads在总Reads中占的比例；</font></br>
<font size=1.5>Reads Mapped Confidently to Transcriptome：比对到已知参考转录本的Reads在总Reads中占的比例；</font></br>
<font size=1.5>Reads Mapped Antisense to Gene：比对到基因反义链上的Reads所占的比例。</font></br>
</details>

### **<font color=black>2 细胞相关统计</font>** {.tabset}
#### Summary
```{r, echo = FALSE}
data <- as.data.frame(t(read.table(file = "metrics_summary.csv",header = T,sep = ",",check.names = F)[,c(1:3,19,18,6,17)]))
data2 <- data.frame(Type = rownames(data),Values = data$V1)
knitr::kable(x = data2,format = 'html',digits = 2,table.attr = "style='width:100%;height:80%'",caption = "**表3 细胞统计**") %>% 
  kableExtra::kable_styling(position = "center",bootstrap_options = c("striped", "hover"),font_size = 16,full_width = T)
```
<details>
<summary>表3 note说明</summary>
<font size=1.5>Estimated Number of Cells：检测到的细胞数；</font></br>
<font size=1.5>Mean Reads per Cell：每个cell中Reads的平均数；</font></br>
<font size=1.5>Median Genes per Cell：每个cell中基因的中位数；</font></br>
<font size=1.5>Median UMI Counts per Cell：每个cell的UMI中位数；</font></br>
<font size=1.5>Total Genes Detected：基因总数；</font></br>
<font size=1.5>Sequencing Saturation：测序饱和度；</font></br>
<font size=1.5>Fraction Reads in Cells：过滤后细胞reads数占总reads数比例。</font></br>
</details>

#### Barcode Rank Plot
```{r, echo=FALSE,warnings = FALSE,fig.show = 'hold',out.width='30%',out.height="30%",force = TRUE}
htmltools::includeHTML(path = "rank_barcode_umi.html")
```

## **<font color=black>3 聚类分析</font>**
```{r, echo=FALSE,fig.show = 'hold',out.width='100%'}
UMAP_df <- read.table(file = "umap_df.xls",header = T,sep = "\t",stringsAsFactors = F)
umap_pic <- ggplot(UMAP_df, aes(x = UMAP_1, y = UMAP_2))+
      geom_point(aes(color = factor(cluster)),size=1, shape=16)+
      scale_color_manual(values = col)+
      labs(colour = "cluster")+
      theme_bw(base_size = 15)+
      theme(panel.grid = element_blank())
      #guides(colour = guide_legend(override.aes = list(size=3.5)))
#ggplotly(umap_pic)

TSNE_df <- read.table(file = "tsne_df.xls",header = T,sep = "\t",stringsAsFactors = F)
tsne_pic <- ggplot(TSNE_df, aes(x = tSNE_1, y = tSNE_2))+
      geom_point(aes(color = factor(cluster)),size=1, shape=16)+
      scale_color_manual(values = col)+
      labs(colour = "cluster")+
      theme_bw(base_size = 15)+
      theme(panel.grid = element_blank())
      #guides(colour = guide_legend(override.aes = list(size=3.5)))
#ggplotly(tsne_pic)
fluidRow(
  column(6,ggplotly(umap_pic) %>% layout(xaxis = list(showgrid = T,showline = T),yaxis = list(showgrid = T,showline = T))),
  column(6,ggplotly(tsne_pic) %>% layout(xaxis = list(showgrid = T,showline = T),yaxis = list(showgrid = T,showline = T)))
)

```
UMAP图形，注重全局，弱化局部结构。在低纬空间保留了高维空间细胞间的关系，反映了细胞群潜在的拓扑结构，对于细胞轨迹推断有很好的的应用。

tSNE图形，注重局部结构，几乎不保留全局信息，可能会夸大细胞群体之间的差异而忽略潜在联系。tSNE倾向于膨胀密集的簇，压缩稀疏的簇。因此不能通过图上大小和位置关系确定细胞群的差异和细胞簇之间的关系。

---

<center> Copyright © 2022-2023 青岛百创智能制造技术有限公司 京ICP备10042835号-6 </center>
<center> 公司地址：山东省青岛市黄岛区保税港区上海路20号三号楼</center>
<center> Tel:400-600-3186 Fax:010-57045001 E-mail:tech@biomarker.com.cn [关于我们](http://www.bmkmanu.com/about) Ver:2.1.2</center>



